text/3724-theupdateframework.md

```diff
@@ -0,0 +1,341 @@
+
+###### Nightly Role
+
+Nightly releases shall be conducted by a single signing key, trusted for only signing nightly releases of the Rust Compiler. This will allow for nightly releases to remain automated and not require the active participation of any rust members.
+
+###### Rustup Role
+
+This shall be a quorum based role, consisting of all members of the Rustup & Infrastructure team members. We recommend having at least a 3-member threshold. We have decided to have this roles quorum be broader to allow for emergency updates and releases of Rustup; we may want to increase the threshold when these teams have more members.
+
+###### Snapshots & Timestamps Role
+
+These roles shall be a single-member role with a key utilized for automation.
+
+### `rust-lang/tuf-crates`
+
+The actual target for tuf-crates shall be the crates index and not the artifacts themselves. This means that the TUF repository for crates.io is performed on much smaller payloads, which still provides us with cryptographic security due to the fact the index contains SHA-256 hashes of the crate file artifacts.  Given the index already consists of SHA-512 signatures of all files, we are then utilizing TUF to validate the index, which in turn is utilized to validate the actual downloaded artifacts. This allows us to perform validation on index updates and not on final downloads, also reducing the overhead of performing multiple hashing and validation procedures on the larger crate artifact files.
+
+#### Roles
+
+##### Root Role
+
+The root role of the tuf-crates repository shall consist of all members of the crates.io rust team with a threshold of 3. As a special case, updating this role shall also require a resigning by the root role of the tuf-root repository (sign a metadata entry existing within tuf-root). This means any changes to the membership of the crates.io team will also require a signing ceremony via github by the root quorum.
+
+##### Targets Role
+
+The repository shall have a top-level Targets role which is utilized directly for all actions on crates.io. This role shall be a 1-key member role, allowing for the automation of actions by crates.io for publishing of crates. In the future, we propose the delegation of roles for trusted publishing within organizations or crates.
+
+###### Snapshots & Timestamps Roles
+
+These roles shall each be a single-member role with a key utilized for automation.
+
+## TUF Management
+
+We propose the adaptation and implementation of TUF-on-CI (https://github.com/theupdateframework/tuf-on-ci) to manage roots and signing events via GitHub CI. This provides a GitHub-centric workflow for performing signing ceremonies via Pull Requests directly on the TUF repositories in question.
+
+Online signing needs shall be implemented with AWS KMS.
+
+All members of all signing quorums within the Rust Project will require hardware keys, the expenses for which will be covered by the Rust Foundation.
+
+## Root Quorum Model
```

Comment:
Some general notes about hardware-backed quorum models, based on our experience on PyPI when attempting to deploy TUF on PyPI (i.e. [PEP 458](https://peps.python.org/pep-0458/)):

1. Hardware (i.e. HSM) backed tokens are difficult to operationalize over expected key lifetimes:
    * You'll next to establish a chain of trust for the hardware itself (some, but not all, HSM models have baked-in device attestations).
    * You'll need to perform a secure offline signing and enrollment ceremony (we [designed a runbook for PyPI](https://github.com/psf/psf-tuf-runbook), but it's pretty old at this point and was also constrainted by HSM vendor limitations that should be re-evaluated)
    * You'll need a *testable* compromise and roation process, to prevent/limit normalization of deviance around key management and enrollment into the quorum.
    * Each quorum party will ideally _physically secure_ their HSM in a way that stymies a medium-complexity physical adversary: PyPI chose tamper-evident bags, and in practice each quorum member will need multiple bags and a tag-in-tag-out procedure for removing their key from their bag if occasional key operations are expected.
1. This section is currently a little light on cryptographic details: it specifies the size of the quorum, but it doesn't say which types and sizes of keys are permitted in the root set, or how the community will verify that a key is actually enrolled _within_ a particular HSM. For PyPI we stipulated a mix of P-256 and P-384 keys due to the limitations of HSMs at the time, but it might be possible to do Ed25519 keypairs with current commercial HSMs. We also prepared HSM-level attestations of key possession, although in practice hardware limitations meant that only the YubiHSMs actually supported root key attestation (versus attestation of the HSM itself).

As a whole, these were nontrivial issues to address in our initial attempt to implement TUF on PyPI, and IMO they're a large part of why TUF (in the form of PEP 458) hasn't materialized on PyPI. So as part of this RFC I recommend the Rust's Leadership Council think about the over/under on a hardware quorum model versus something simpler (e.g. a soft root key in a cloud HSM) or even a different, smaller-footprint architecture (like a transparency scheme). I'll leave a separate comment on the latter.

Reply:
To add to what @woodruffw wrote, for our HSM enrolment we require each key-holder to upload a device and key attestation that can be verified up to the manufacturer's root CA. Verification of this is done via GitHub actions, so for each new key added we get automatic verification. 

We also hook this into the TUF verification process, so that each time a TUF metadata document is updated we run a workflow that verify the TUF signatures via HSM's attested device key. This may be slightly overkill, but makes it simple and foolproof to see that no extra keys has been added that is not known and approved.